src/common/eigf.h

```diff
@@ -0,0 +1,229 @@
+ * - average of mask
+ * - covariance of mask and guide. */
+static inline void eigf_variance_analysis(const float *const restrict guide, // I
+                                    const float *const restrict mask, //p
+                                    float *const restrict out,
+                                    const size_t width, const size_t height,
+                                    const float sigma)
+{
+  // We also use gaussian blurs instead of the square blurs of the guided filter
+  const size_t Ndim = width * height;
+  float *const restrict in = dt_alloc_sse_ps(Ndim * 4);
+
+  float ming = 10000000.0f;
+  float maxg = 0.0f;
+  float minm = 10000000.0f;
+  float maxm = 0.0f;
+  float ming2 = 10000000.0f;
+  float maxg2 = 0.0f;
+  float minmg = 10000000.0f;
+  float maxmg = 0.0f;
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+dt_omp_firstprivate(guide, mask, in, Ndim) \
+  schedule(simd:static) aligned(guide, mask, in:64) \
+  reduction(max:maxg, maxm, maxg2, maxmg)\
+  reduction(min:ming, minm, ming2, minmg)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    const float pixelg = guide[k];
+    const float pixelm = mask[k];
+    const float pixelg2 = pixelg * pixelg;
+    const float pixelmg = pixelm * pixelg;
+    in[k * 4] = pixelg;
+    in[k * 4 + 1] = pixelg2;
+    in[k * 4 + 2] = pixelm;
+    in[k * 4 + 3] = pixelmg;
+    if(pixelg < ming) ming = pixelg;
+    if(pixelg > maxg) maxg = pixelg;
+    if(pixelm < minm) minm = pixelm;
+    if(pixelm > maxm) maxm = pixelm;
+    if(pixelg2 < ming2) ming2 = pixelg2;
+    if(pixelg2 > maxg2) maxg2 = pixelg2;
+    if(pixelmg < minmg) minmg = pixelmg;
+    if(pixelmg > maxmg) maxmg = pixelmg;
+  }
+
+  float max[4] = {maxg, maxg2, maxm, maxmg};
+  float min[4] = {ming, ming2, minm, minmg};
+  dt_gaussian_t *g = dt_gaussian_init(width, height, 4, max, min, sigma, 0);
+  if(!g) return;
+  dt_gaussian_blur_4c(g, in, out);
+  dt_gaussian_free(g);
+
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+dt_omp_firstprivate(out, Ndim) \
+  schedule(simd:static) aligned(out:64)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    out[4 * k + 1] -= out[4 * k] * out[4 * k];
+    out[4 * k + 3] -= out[4 * k] * out[4 * k + 2];
+  }
+
+  dt_free_align(in);
+}
+
+void eigf_blending(float *const restrict image, const float *const restrict av,
+                  const size_t Ndim, const dt_iop_guided_filter_blending_t filter,
+                  const float feathering)
+{
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(image, av, Ndim, feathering, filter) \
+  schedule(simd:static) aligned(image, av:64)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    const float avg_g = av[k * 4];
+    const float avg_m = av[k * 4 + 2];
+    const float var_g = av[k * 4 + 1];
+    const float covar_mg = av[k * 4 + 3];
+    const float norm_g = fmaxf(avg_g * image[k], 1E-6);
+    const float norm_m = fmaxf(avg_m * image[k], 1E-6);
```

Comment:
shouldn't this be `avg_m * mask[k]` ?

Reply:
mask is computed on downscaled image for performance optimisation, so does not exist at full resolution, that's why I use image[k]

Reply:
But we can quantize on full resolution and downscale the mask if you prefer

Reply:
I think it would make sense to use the mask value here, just from a theoretical point of view.

Reply:
I agree, I fixed that in my last push. The change in performance seems very small :-)